### PR TITLE
begin separation of x509 crate from cryptography crate

### PR DESCRIPTION
**Pull request by reaperhulk**
this will not be a published crate for now and the separation is incomplete. Putting up for commentary, although if we agree this is generally the right path we'll want to land this (+ review feedback) and then start tracking some tasks for improvement. The goal here is a crate that we may choose to publish that is useful to rust users and provides a clean separation of x509 parse (and encode?) logic from our pyo3 code.

### CODE REVIEW

File: src/rust/Cargo.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ rust-version = "1.56.0"
 once_cell = "1"
 pyo3 = { version = "0.18" }
 asn1 = { version = "0.14.0", default-features = false }
+cryptography-x509 = { version = "0.1.0", path = "../rust_x509" }
```
Comment (alex) [inline src/rust/Cargo.toml:14]: Is `version` required?

File: src/rust/src/pkcs7.rs [hunk 2]
```diff
@@ -6,6 +6,10 @@ use crate::asn1::encode_der_data;
 use crate::buf::CffiBuf;
 use crate::error::CryptographyResult;
 use crate::x509;
+use cryptography_x509::certificate::RawCertificate;
```
Comment (alex) [inline src/rust/src/pkcs7.rs:9]: I don't think it should be named RawCertificate in that crate?

File: src/rust/src/x509/certificate.rs [hunk 3]
```diff
@@ -767,21 +613,14 @@ pub(crate) fn encode_distribution_point_reasons(
     Ok(asn1::OwnedBitString::new(bits, unused_bits).unwrap())
 }
 
+// TODO: not moved since it has a FromPyObject on it
```
Comment (alex) [inline src/rust/src/x509/certificate.rs:616]: 🙃

File: src/rust/src/x509/mod.rs [hunk 4]
```diff
@@ -10,13 +10,13 @@ pub(crate) mod extensions;
...
 
 pub(crate) use certificate::Certificate;
 pub(crate) use common::{
     datetime_to_py, find_in_pem, parse_and_cache_extensions, parse_general_name,
-    parse_general_names, parse_name, parse_rdn, py_to_datetime, AlgorithmIdentifier,
-    Asn1ReadableOrWritable, AttributeTypeValue, Extensions, GeneralName, Name, Time,
+    parse_general_names, parse_name, parse_rdn, py_to_datetime,
 };
+pub(crate) use cryptography_x509::common::{AlgorithmIdentifier, Asn1ReadableOrWritable, Time};
+pub(crate) use cryptography_x509::name::{GeneralName, Name};
```
Comment (alex) [inline src/rust/src/x509/mod.rs:22]: I don't think there's a good reason to re-export these names here? Stuff should just import from teh right place

File: src/rust_x509/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[package]
```
Comment (alex) [inline src/rust_x509/Cargo.toml:1]: This should not have it's own Cargo.lock. Look into workspaces.

File: src/rust_x509/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+name = "cryptography-x509"
```
Comment (alex) [inline src/rust_x509/Cargo.toml:2]: Is there a reason not make this match the directory name?

File: src/rust_x509/src/certificate.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+use crate::common;
+use crate::extensions;
+use crate::name;
+use asn1;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write, Hash, PartialEq, Clone)]
+pub struct RawCertificate<'a> {
```
Comment (alex) [inline src/rust_x509/src/certificate.rs:7]: Why Raw?

File: src/rust_x509/src/certificate.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+use crate::common;
+use crate::extensions;
+use crate::name;
+use asn1;
```
Comment (alex) [inline src/rust_x509/src/certificate.rs:4]: Not necessary.

File: src/rust_x509/src/crl.rs [hunk 9]
```diff
@@ -0,0 +1,66 @@
+use crate::{common, extensions, name};
+use asn1;
+
+pub type ReasonFlags<'a> =
+    Option<common::Asn1ReadableOrWritable<'a, asn1::BitString<'a>, asn1::OwnedBitString>>;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write, PartialEq, Hash)]
+pub struct RawCertificateRevocationList<'a> {
```
Comment (alex) [inline src/rust_x509/src/crl.rs:8]: Same

File: src/rust_x509/src/crl.rs [hunk 10]
```diff
@@ -0,0 +1,66 @@
+use crate::{common, extensions, name};
+use asn1;
```
Comment (alex) [inline src/rust_x509/src/crl.rs:2]: Same

File: src/rust_x509/src/crl.rs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    pub issuer: name::Name<'a>,
+    pub this_update: common::Time,
+    pub next_update: Option<common::Time>,
+    pub revoked_certificates: RevokedCertificates<'a>,
+    #[explicit(0)]
+    pub crl_extensions: Option<extensions::Extensions<'a>>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write, PartialEq, Hash, Clone)]
+pub struct RawRevokedCertificate<'a> {
```
Comment (alex) [inline src/rust_x509/src/crl.rs:35]: Same :-)

File: src/rust_x509/src/csr.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
+use crate::common;
+use crate::extensions;
+use crate::name;
+use crate::oid;
+use asn1;
+
+// TODO: rename "raw"
```
Comment (alex) [inline src/rust_x509/src/csr.rs:7]: Seeee!

**Review by alex — COMMENTED** [review]
Plz don't rebase for ease of future reviews :-)

Does the coverage + cargo test stuff not need updating? (I think the cargo test stuff will just work if you set up the workspace)

**Comment by reaperhulk** [comment]
I've fixed coverage mostly -- there's one missed line on a derive statement that I don't understand but the other is probably real.

**Comment by reaperhulk** [comment]
One issue that I don't understand but otherwise full coverage now.

File: src/rust/cryptography-x509/src/extensions.rs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    'a,
+    asn1::SequenceOf<'a, Extension<'a>>,
+    asn1::SequenceOfWriter<'a, Extension<'a>, Vec<Extension<'a>>>,
+>;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write, PartialEq, Hash, Clone)]
+pub struct AlgorithmIdentifier<'a> {
+    pub oid: asn1::ObjectIdentifier,
+    pub params: Option<asn1::Tlv<'a>>,
+}
```
Comment (alex) [inline src/rust/cryptography-x509/src/extensions.rs:19]: This is a duplicate of this same type defined elsewhere, it's unused.

**Review by alex — APPROVED** [review]

**Pull request merged**